### PR TITLE
Initial V3 Migration Guide

### LINKED ISSUE #251: Create migration guide for version 3
The following changes must be covered:

- #249
- #130 
- #254
- _must be continued…_

### LINKED ISSUE #292: Add example code to the migration guide for dealing with final validators
See #255

https://github.com/laminas/laminas-validator/pull/255#issuecomment-2188440340

Migration guide patch: #253

### PR DESCRIPTION
**Pull request by gsteel**
- Closes #251 
- Closes #292

### CODE REVIEW

**Comment by gsteel** [comment]
@froschdesign - Could you please cast your eyes over this to see if I'm going in the right direction. It's getting pretty big and I'd appreciate a bit of feedback before I put too much more effort into it in case I'm doing a terrible job!!

File: mkdocs.yml [hunk 1]
```diff
@@ -58,6 +58,8 @@ nav:
             - Size: v3/validators/file/size.md
             - UploadFile: v3/validators/file/upload-file.md
             - WordCount: v3/validators/file/word-count.md
+        - Migration:
+            - "Migration from Version 2 to 3": v3/migration/v2-to-v3.md
```
Comment (froschdesign) [inline mkdocs.yml:65]: The new page "Refactoring Legacy Validators" is a great idea and of course the `MuppetNameValidator`! 👍🏻 😄 
But this page must also be included in the navigation.

**Comment by froschdesign** [comment]
@gsteel 
> Could you please cast your eyes over this to see if I'm going in the right direction.

I must apologise, because of course I've looked at every change to this pull request so far, but left you in the dark about it.
The direction is perfect! 👍🏻 
I will do another check to make sure that nothing has been forgotten.

**Comment by gsteel** [comment]
> I must apologise, because of course I've looked at every change to this pull request so far, but left you in the dark about it. The direction is perfect! 👍🏻 I will do another check to make sure that nothing has been forgotten.

Thank you @froschdesign 🙏

**Comment by froschdesign** [comment]
@gsteel 
Since the question will come up, I'm thinking about what to write to the support of laminas-mvc with version 3?!
The other topic is the database validators. At the moment we are not saying anything about alternatives or offering anything else. 🤔

**Comment by gsteel** [comment]
For the DB validators, Given that `laminas-db` is unmaintained, perhaps we should point people in the direction of [Doctrine Module](https://github.com/doctrine/DoctrineModule/tree/6.2.x/src/Validator).

If MVC:3 were to receive ServiceManager:4 support, then Validator:3 is theoretically compatible, once Filter, InputFilter, and Form are.

I think we should just state that Validator:3 is not compatible with MVC until SM:4 support is available there - Whether that appears in MVC:3 or MVC:4 is then not Validators' concern (??)

**Comment by froschdesign** [comment]
@gsteel 
I will think of something and then write about it. You don't have to do that! 😃

**Comment by gsteel** [comment]
Thanks @froschdesign - I was just actually writing something about DB - This is what I cam up with, but I'll not commit it and leave this up to you:

### `Laminas\Db` Validator Removal

The deprecated "Db" validators that shipped in version 2.0 have been removed. The removed classes are:

- `Laminas\Validator\Db\AbstractDb`
- `Laminas\Validator\Db\NoRecordExists`
- `Laminas\Validator\Db\RecordExists`

There are now no maintained validators that work directly with the `laminas-db` component.

As `laminas-db` has been in [security-only maintenance mode](https://github.com/laminas/technical-steering-committee/blob/main/meetings/minutes/2022-02-07-TSC-Minutes.md) for quite some time, we recommend that users consider maintained alternatives such as [Doctrine DBAL](https://www.doctrine-project.org/projects/dbal.html).

The [Doctrine Project](https://www.doctrine-project.org) provides [Doctrine Module](https://www.doctrine-project.org/projects/doctrine-module.html) to integrate with the Laminas ecosystem.

Alternatively, you can extract the `laminas-db` related validators from the most recent `2.x` release, refactor them to be compatible with version 3 of laminas-validator and incorporate them into your own project.

**Comment by froschdesign** [comment]
@gsteel 
Do you think it would be a problem to add the [`Laminas\Db\Adapter\AdapterServiceDelegator`](https://docs.laminas.dev/laminas-db/adapters/adapter-aware-trait/#adapterservicedelegator) for database validators for version 2 in a minor version?
This would allow us to refer to simple use and implementation.

**Comment by gsteel** [comment]
Do you mean automatically register the delegator in ConfigProvider and Module?

I have no problem with that - I'm surprised it wasn't done before looking at the docs that you linked!

**Comment by froschdesign** [comment]
My current idea:

- add the delegator for the database validators (v2)
- add the usage of validators in a laminas-mvc based application to the documentation (v2)
- add the usage of database validators in forms in a laminas-mvc based application to documentation (v2)
- add links to application integrations to homepage (v2) - [example](https://docs.laminas.dev/laminas-eventmanager/#learn)
- add a warning when using laminas-mvc to homepage (v3)

**Comment by gsteel** [comment]
Sounds like a good plan to me, however, it's a fair bit of work for legacy stuff that people should be migrating away from, even though it will be a clear benefit to users. Is it worth the effort?

**Comment by froschdesign** [comment]
The point is quite right, but we don't offer an alternative and then also say that version 3 doesn't work with laminas-mvc.

**Comment by gsteel** [comment]
@froschdesign Can this be merged as is? And, can we amend docs post 3.0 release? I'm keen to get 3.0 out the door to enable work on dependents. Are there any other blockers I'm not considering?

**Comment by froschdesign** [comment]
@gsteel 
> Can this be merged as is? And, can we amend docs post 3.0 release?

Sure, go for it!

**Pull request merged**